### PR TITLE
V0.50.1 patched

### PR DESCRIPTION
**Pull request by yan-soon**
key changes:
- revert new cosmos-sdk change where `multiSend` can only have one sender (https://github.com/cosmos/cosmos-sdk/issues/12601)
- revert new cosmos-sdk change where `send` and `multiSend` typeAsserts the bank keeper to be of type `BaseKeeper` (this type assertion has issues with our implementation as our bank keeper is customised due to the x/alliance module, where the keeper is a wrapper of the original `x/bank BaseKeeper`, resulting in the type assertion check failing 
https://github.com/Switcheo/cosmos-sdk/pull/438/commits/102b3f27ed74d6b62851c91c6444c887a872d6a9

### CODE REVIEW

**Comment by RavenXce** [comment]
reverting msgmultisend is not safe

File: types/module/module.go [hunk 1]
```diff
@@ -777,6 +792,8 @@ func (m *Manager) BeginBlock(ctx sdk.Context) (sdk.BeginBlock, error) {
 	for _, moduleName := range m.OrderBeginBlockers {
 		if module, ok := m.Modules[moduleName].(appmodule.HasBeginBlocker); ok {
 			if err := module.BeginBlock(ctx); err != nil {
+				startTime := time.Now()
```
Comment (RavenXce) [inline types/module/module.go:795]: huh is start here correct..?

File: types/module/module.go [hunk 2]
```diff
@@ -796,15 +813,29 @@ func (m *Manager) EndBlock(ctx sdk.Context) (sdk.EndBlock, error) {
 
 	for _, moduleName := range m.OrderEndBlockers {
 		if module, ok := m.Modules[moduleName].(appmodule.HasEndBlocker); ok {
+			if m.beforeModuleEndBlock != nil {
+				m.beforeModuleEndBlock(moduleName)
```
Comment (RavenXce) [inline types/module/module.go:817]: no end block telemetry..?

File: types/errors/abci.go [hunk 3]
```diff
@@ -30,7 +30,6 @@ func ResponseExecTxResultWithEvents(err error, gw, gu uint64, events []abci.Even
 		Log:       log,
 		GasWanted: int64(gw),
 		GasUsed:   int64(gu),
-		Events:    events,
```
Comment (RavenXce) [inline types/errors/abci.go:33]: does this change things for us?

File: store/snapshots/store.go [hunk 4]
```diff
@@ -274,6 +274,12 @@ func (s *Store) Save(
 
 			dirCreated = true
 		}
+		path := s.PathChunk(height, format, index)
+		file, err := os.Create(path)
+		if err != nil {
+			return nil, sdkerrors.Wrapf(err, "failed to create snapshot chunk file %q", path)
+		}
```
Comment (RavenXce) [inline store/snapshots/store.go:281]: what patch is this?

File: store/pruning/types/options.go [hunk 5]
```diff
@@ -16,16 +16,20 @@ type PruningOptions struct {
 
 	// Strategy defines the kind of pruning strategy. See below for more information on each.
 	Strategy PruningStrategy
+
+	// PruningStartHeight defines the start height to prune from.
+	PruningStartHeight uint64
```
Comment (RavenXce) [inline store/pruning/types/options.go:21]: what patch is this?

File: store/pruning/types/options.go [hunk 6]
```diff
@@ -16,16 +16,20 @@ type PruningOptions struct {
...
 const (
-	PruningOptionDefault    = "default"
-	PruningOptionEverything = "everything"
-	PruningOptionNothing    = "nothing"
-	PruningOptionCustom     = "custom"
+	PruningOptionDefault     = "default"
+	PruningOptionEverything  = "everything"
+	PruningOptionNothing     = "nothing"
+	PruningOptionCustom      = "custom"
+	PruningOptionStartHeight = "pruning-start-height"
```
Comment (RavenXce) [inline store/pruning/types/options.go:32]: what patch is this?

File: server/config/toml.go [hunk 7]
```diff
@@ -136,6 +140,9 @@ swagger = {{ .API.Swagger }}
 # Address defines the API server to listen on.
 address = "{{ .API.Address }}"
 
+# Oracle-address defines the API server to listen on for oracle service.
+oracle-address = "{{ .API.OracleAddress }}"
```
Comment (RavenXce) [inline server/config/toml.go:144]: y are there oracle stuff here?

File: server/config/config.go [hunk 8]
```diff
@@ -241,13 +247,15 @@ func DefaultConfig() *Config {
...
-			GlobalLabels: [][]string{},
+			Enabled:                  false,
+			GlobalLabels:             [][]string{},
+			MonitorStoreSizeInterval: 0,
 		},
 		API: APIConfig{
 			Enable:             false,
 			Swagger:            false,
 			Address:            DefaultAPIAddress,
+			OracleAddress:      DefaultOracleAPIAddress,
```
Comment (RavenXce) [inline server/config/config.go:258]: y are there oracle stuff here?

File: server/config/config.go [hunk 9]
```diff
@@ -108,6 +111,9 @@ type APIConfig struct {
 	// Address defines the API server to listen on
 	Address string `mapstructure:"address"`
 
+	// OracleAddress defines if the API server to listen on for oracle service
+	OracleAddress string `mapstructure:"oracle-address"`
```
Comment (RavenXce) [inline server/config/config.go:115]: y are there oracle stuff here?

**Review by RavenXce — CHANGES_REQUESTED** [review]

File: server/config/toml.go (hunk 7 above)
Comment (yan-soon) [reply server/config/toml.go:144]: https://github.com/yan-soon/cosmos-sdk/commit/48e32471b66a86a806e614e320af1df6402d7398

File: server/config/config.go (hunk 8 above)
Comment (yan-soon) [reply server/config/config.go:258]: https://github.com/switcheo/cosmos-sdk/commit/48e32471b66a86a806e614e320af1df6402d7398

File: server/config/config.go (hunk 9 above)
Comment (yan-soon) [reply server/config/config.go:115]: https://github.com/switcheo/cosmos-sdk/commit/48e32471b66a86a806e614e320af1df6402d7398

File: baseapp/abci.go [hunk 10]
```diff
@@ -686,6 +686,103 @@ func (app *BaseApp) VerifyVoteExtension(req *abci.RequestVerifyVoteExtension) (r
 	return resp, err
 }
 
+// BeginBlock implements the ABCI application interface.
+// kept just for testing purposes on carbon side
+func (app *BaseApp) BeginBlock(req *abci.RequestFinalizeBlock) (res sdk.BeginBlock, err error) {
```
[the remark this answers is not in the document]
Comment (yan-soon) [reply baseapp/abci.go]: resolved in https://github.com/Switcheo/cosmos-sdk/pull/438/commits/2bb73567da711a92545f5b598e06bc8c37d5d18d

File: x/evidence/abci.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+	abci "github.com/cometbft/cometbft/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/evidence/keeper"
+	"github.com/cosmos/cosmos-sdk/x/evidence/types"
+)
+
+// BeginBlocker iterates through and handles any newly discovered evidence of
+// misbehavior submitted by Tendermint. Currently, only equivocation is handled.
+func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) {
```
[the remark this answers is not in the document]
Comment (yan-soon) [reply x/evidence/abci.go]: resolved in https://github.com/Switcheo/cosmos-sdk/pull/438/commits/2bb73567da711a92545f5b598e06bc8c37d5d18d

File: x/distribution/types/params.go [hunk 12]
```diff
@@ -40,6 +41,25 @@ func validateCommunityTax(i interface{}) error {
 	return nil
 }
 
+func validateLiquidityProviderReward(i interface{}) error {
```
[the remark this answers is not in the document]
Comment (yan-soon) [reply x/distribution/types/params.go:53]: resolved in https://github.com/Switcheo/cosmos-sdk/pull/438/commits/2bb73567da711a92545f5b598e06bc8c37d5d18d

File: x/bank/keeper/msg_server.go [hunk 13]
```diff
@@ -32,17 +32,31 @@ func (k msgServer) Send(goCtx context.Context, msg *types.MsgSend) (*types.MsgSe
...
-		to, err = base.ak.AddressCodec().StringToBytes(msg.ToAddress)
-		if err != nil {
-			return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid to address: %s", err)
-		}
-	} else {
-		return nil, sdkerrors.ErrInvalidRequest.Wrapf("invalid keeper type: %T", k.Keeper)
+	// original cosmos-sdk implementation, using custom implementation due to type assertion issues
+	// with custom wrapper alliance bank module
+
+	// if base, ok := k.Keeper.(*BaseKeeper); ok {
```
[the remark this answers is not in the document]
Comment (yan-soon) [reply x/bank/keeper/msg_server.go:38]: the type assertion check : `if base, ok := k.Keeper.(*BaseKeeper); ok {`
of the custom wrapper alliance bank module to `*BaseKeeper` fails

**Pull request closed without merging**